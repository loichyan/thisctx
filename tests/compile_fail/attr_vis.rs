// Arbitray tokens in custom attributes are supported since Rust 1.34.
// https://blog.rust-lang.org/2019/04/11/Rust-1.34.0.html#custom-attributes-accept-arbitrary-token-streams
mod error {
    use thisctx::WithContext;

    #[derive(WithContext)]
    #[thisctx(vis = "pub(crate)")]
    pub enum Error {
        #[thisctx(vis = "pub")]
        PubVariant(i32),
        PubCrateVariant(i32),
        #[thisctx(vis = "")]
        PrivateVariant(i32),
        PrivateField(#[thisctx(vis = "")] i32),
    }
}

pub use error::PrivateVariant;
pub use error::PubCrateVariant;
pub use error::PubVariant;

fn any<T>() -> T {
    todo!()
}

fn main() {
    let t = any::<error::PrivateField>();
    t.0;
}
