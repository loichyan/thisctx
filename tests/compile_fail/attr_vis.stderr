error[E0364]: `PubCrate` is private, and cannot be re-exported
  --> tests/compile_fail/attr_vis.rs:23:9
   |
23 | pub use error::PubCrate as _;
   |         ^^^^^^^^^^^^^^^^^^^^
   |
note: consider marking `PubCrate` as `pub` in the imported module
  --> tests/compile_fail/attr_vis.rs:23:9
   |
23 | pub use error::PubCrate as _;
   |         ^^^^^^^^^^^^^^^^^^^^

error[E0603]: struct `Private` is private
  --> tests/compile_fail/attr_vis.rs:33:12
   |
33 | use error::Private as _;
   |            ^^^^^^^ private struct
   |
note: the struct `Private` is defined here
  --> tests/compile_fail/attr_vis.rs:7:14
   |
7  |     #[derive(thisctx::WithContext)]
   |              ^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the derive macro `thisctx::WithContext` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0603]: tuple struct constructor `Private` is private
  --> tests/compile_fail/attr_vis.rs:39:20
   |
16 |         Private(i32),
   |                 --- a constructor is private if any of the fields is private
...
39 |     let _ = error::Private(0);
   |                    ^^^^^^^ private tuple struct constructor
   |
note: the tuple struct constructor `Private` is defined here
  --> tests/compile_fail/attr_vis.rs:7:14
   |
7  |     #[derive(thisctx::WithContext)]
   |              ^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the derive macro `thisctx::WithContext` (in Nightly builds, run with -Z macro-backtrace for more info)
